### ISSUE TITLE
Cloud Spanner Client should prevent nested transactions

### ISSUE THREAD

**Issue by vkedia** [GoogleCloudPlatform/google-cloud-ruby#1519]
Cloud Spanner does not support [nested transactions](https://en.wikipedia.org/wiki/Nested_transaction). But the client library does allow users to call `database.transaction` inside of the callback that is provided to an outer `database.transaction`. This is misleading since users might believe that this will behave like a nested transaction but in reality these will be two independent transactions. This is confusing and a source of bugs. Specifically the inner transaction might succeed while the outer might ABORT, in which case the callback will be rerun thus rerunning the inner transaction. 
We should prevent this from happening by detecting and raising an error if someone calls `database.transaction` inside the callback provided to `database.transaction`.  Note that this needs to be done before Beta since this is a known breaking change.
Also note that if we come across a use case where we do want to allow calling a nested `database.transaction`, we should be able to enable that in future without making a breaking change. 

cc @jgeewax @bjwatson

**Comment by blowmage** [comment]
How do other languages detect and prevent this?

**Comment by blowmage** [comment]
I expect we can put the transaction_id on the current Thread when a transaction starts, and check for the value when other transactions start, but that may block users from creating and using concurrent transactions with Fibers.

I suppose we could walk the closure's AST and detect if it opens another transaction, but I don't know if that approach is performant enough for production use. This is an interesting problem...

**Comment by vkedia** [comment]
Does Ruby have Fibers? Where can I read up more on them? Ideally as you said we would put something on Thread local but presence of FIbers make this interesting. 
In Go, which has goroutines which are in some sense similar to Fibers, there is a Context object being passed around. So we use that to record the fact that there is already an ongoing transaction.

**Comment by blowmage** [comment]
Yes, Ruby does have a [`Fiber`](https://ruby-doc.org/core-2.0.0/Fiber.html) class.

https://ruby-doc.org/core-2.0.0/Fiber.html

**Comment by vkedia** [comment]
I guess we can use Fiber local variables
https://ruby-doc.org/core-2.2.2/Thread.html#class-Thread-label-Fiber-local+vs.+Thread-local

**Comment by blowmage** [comment]
Yeah, it looks like we get/set values using `Thread.current[]` using either threads or fibers and the behavior will be as expected.

That is, unless users create a nested thread or fiber in the transaction block:

```ruby
client.transaction do |tx|
  tx.insert "users", [{ id: 2, name: "Harvey",  active: true }]

  Thread.new do
    # This nested transaction would be allowed
    # since it is happening on a different thread.
    client.transaction do |tx2|
      tx2.insert "users", [{ id: 6, name: "Barney",  active: true }]
    end
  end.join
end
```

**Comment by vkedia** [comment]
I think we are ok allowing that case. The intent here is to prevent people
from mistakenly use nested transactions and expect them to behave in a
certain way.

**Comment by blowmage** [comment]
Okay, so does #1522 satisfy this request?

**Comment by bjwatson** [comment]
@vkedia Can we get a review of https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/1522? Thanks!

**Comment by blowmage** [comment]
Closed via #1522.
